### PR TITLE
[WIP][stdlib] Minor Sequence Optimizations

### PR DESCRIPTION
**Pull request by palimondo**
Following up on discussion from https://github.com/apple/swift/pull/20221#pullrequestreview-174666315, these are minor tweaks to `Sequence` implementation:
* In-place rotation of `ringBuffer` used in `suffix`, avoiding new array copy.
* Internal use of `ContiguousArray` in `dropLast`, `prefix` and `suffix` (modeled after `map` and `_filter` implementations).

### CODE REVIEW

**Comment by palimondo** [comment]
When measuring locally, I saw no performance difference. But I don't think our benchmarks, slicing only short sequences of Int`s, cover the affected area very thoroughly…

@airspeedswift Can this be accepted only on theoretical memory-saving grounds? Or do we need to devise better benchmark that showcases the tradeoffs between in-place rotation and array copy?

**Comment by palimondo** [comment]
Hmm… I don't recall seeing any regressions with SuffixSequence when I ran this locally 2 weeks ago. 🤨

**Comment by palimondo** [comment]
@airspeedswift Any idea why that ^^ might be?

**Comment by palimondo** [comment]
Let's try again…
@swift-ci please benchmark

**Comment by palimondo** [comment]
I cannot reproduce the regression on `SuffixSequence` locally. Let's try remote performance debugging… First, reverting all changes to suffix for a clean baseline.

@swift-ci please benchmark

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@airspeedswift I think I have this down now… can you please have a look, because in order to work around the weird de-optimization in `SuffixSequence`, I had to take the array rotation out of the `suffix` method itself. What would be a proper way to declare a non-public utility extension like I did here for `ContiguousArray.rotate(left:)` inside the stdlib? I can't quickly find other precedent.

**Comment by palimondo** [comment]
CC @dabrahams @lorentey

File: stdlib/public/core/Sequence.swift [hunk 1]
```diff
@@ -800,6 +800,18 @@ extension Sequence where Element : Equatable {
   }
 }
 
+extension ContiguousArray {
+  /// Rotates array in place to the left by specified distance.
+  @usableFromInline
+  mutating func _rotate(left distance: Int) {
+    guard distance > 0 else { return }
+    let i = index(startIndex, offsetBy: distance)
```
Comment (dabrahams) [inline stdlib/public/core/Sequence.swift]: This is not necessarily the best way to rotate, and is probably not the API we want to commit to supporting even internally.  Please try to use what's already written in the algorithms prototype, and especially read [this comment](https://github.com/apple/swift/blob/bf1df4fdf8881d0ff50a552b99c7ffb3afad66b7/test/Prototypes/Algorithms.swift#L193-L203) and test accordingly.

Comment (palimondo) [reply stdlib/public/core/Sequence.swift]: I guess it's better than to split this and do just the consistent use of `ContiguousArray` in this PR, and tackle the in-place rotation in a separate PR, right?

**Pull request closed without merging**